### PR TITLE
Rational - Cache field data to sqlite

### PR DESCRIPTION
**Pull request by sibartlett**

### CODE REVIEW

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 1]
```diff
@@ -201,6 +218,12 @@ def _rational_forms_login(self, forms_client):
             raise AlmException('Unable to connect to %s (Check server URL, '
                                'user, pass). Reason: %s' % (self.alm_name, str(err)))
 
+    def shorten_alm_url(self, url):
+        return url.replace(self.alm_plugin.base_uri + '/', '')
```
Comment (hanifvirani) [inline sdetools/modules/sync_rational/rational_plugin.py]: Use ```.format ``` here.

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 2]
```diff
@@ -201,6 +218,12 @@ def _rational_forms_login(self, forms_client):
...
+
+    def unshorten_alm_url(self, url):
+        return '%s/%s' % (self.alm_plugin.base_uri, url)
```
Comment (hanifvirani) [inline sdetools/modules/sync_rational/rational_plugin.py]: Same.

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 3]
```diff
@@ -292,32 +315,37 @@ def _get_rtc_field(self, field_name, field_value):
...
-
-            if value_name == field_value:
-                return {
-                    'name': field_name,
-                    'value': {
-                        'rdf:resource': allowed_value['rdf:resource']
-                    }
-                }
+        if 'oslc:allowedValues' in rs:
+            return self._get_rtc_restricted_value(field_name, field_value, rs)
```
Comment (hanifvirani) [inline sdetools/modules/sync_rational/rational_plugin.py]: ```rs``` seems like a bad name here. Doesn't tell me what it is on reading.

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 4]
```diff
@@ -292,32 +315,37 @@ def _get_rtc_field(self, field_name, field_value):
...
+
+        return RationalField(field_name, field_value)
+
+    def _get_rtc_restricted_value(self, field_name, field_value, rs):
+        resource_urls = [
+            allowed_value['rdf:resource']
+            for allowed_value in rs['oslc:allowedValues']['oslc:allowedValue']
+        ]
+
+        cached_value = self.field_cache[field_name].get(field_value, None)
```
Comment (hanifvirani) [inline sdetools/modules/sync_rational/rational_plugin.py]: Can this result in key errors?

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 5]
```diff
@@ -292,32 +315,37 @@ def _get_rtc_field(self, field_name, field_value):
...
-                    'value': {
-                        'rdf:resource': allowed_value['rdf:resource']
-                    }
-                }
+        if 'oslc:allowedValues' in rs:
+            return self._get_rtc_restricted_value(field_name, field_value, rs)
+
+        return RationalField(field_name, field_value)
+
+    def _get_rtc_restricted_value(self, field_name, field_value, rs):
```
Comment (hanifvirani) [inline sdetools/modules/sync_rational/rational_plugin.py]: Same here with ```rs```

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 6]
```diff
@@ -431,7 +459,7 @@ def alm_add_task(self, task):
         for field_name, field_value in new_artifact.iteritems():
             field = self._get_rtc_field(field_name, field_value)
             if field:
-                create_args[field['name']] = field['value']
+                create_args[field.name] = field.value
 
         if self.config['alm_standard_workflow'] and (task['status']['id'] in STATUS.DONE_SET or task['status']['id'] in STATUS.NA_SET):
```
Comment (hanifvirani) [inline sdetools/modules/sync_rational/rational_plugin.py]: Loong line.

File: sdetools/extlib/sqlitedict.py [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+try:
+    from queue import Queue
+except ImportError:
+    from Queue import Queue
+
+
+logger = logging.getLogger(__name__)
+
+
+def open(*args, **kwargs):
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Is it weird to override a python keyword like this?

File: sdetools/sdelib/sqlite.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+from sdetools.sdelib import log_mgr
+
+
+logger = log_mgr.mods.add_mod(__name__)
+
+
+class SqliteDictManager(collections.defaultdict):
+    """
+    Manages multiple instances of SqliteDict
+    """
```
Comment (hanifvirani) [inline sdetools/sdelib/sqlite.py]: Blank line after class docstring

File: sdetools/extlib/sqlitedict.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+        The `flag` parameter. Exactly one of:
+          'c': default mode, open for read/write, creating the db/table if necessary.
+          'w': open for r/w, but drop `tablename` contents first (start with empty table)
+          'r': open as read-only
+          'n': create a new database (erasing any existing tables, not just `tablename`!).
+
+        """
+        self.in_temp = filename is None
+        if self.in_temp:
+            randpart = hex(random.randint(0, 0xffffff))[2:]
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Doesn't python have code for giving you a temp file as well as a temp folder? Or do you want the name to be sqldictBLAH?

File: sdetools/extlib/sqlitedict.py [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+            filename = os.path.join(tempfile.gettempdir(), 'sqldict' + randpart)
+
+        if flag not in SqliteDict.VALID_FLAGS:
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Do this check first so you don't bother making the temporary file if it'll fail.

File: sdetools/extlib/sqlitedict.py [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+            raise RuntimeError("Unrecognized flag: %s" % flag)
+        self.flag = flag
+
+        if flag == 'n':
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]:
```
if flag == 'n' and os.path.exists(filename):
    os.remove(filename)
```

File: sdetools/extlib/sqlitedict.py [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+            if os.path.exists(filename):
+                os.remove(filename)
+
+        dirname = os.path.dirname(filename)
+        if dirname:
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]:
```
if dirname and not os.path.exists(dirname):
    raise RuntimeError('Error! The directory does not exist, %s' % dirname)
```

File: sdetools/sdelib/sqlite.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        return SqliteDict(filename, autocommit=True)
+
+    def ensure_directory_exists(self):
+        if not os.path.exists(self.directory):
+            logger.debug('Creating database directory: %s' % self.directory)
+            os.makedirs(self.directory)
+
+    def get_filename(self, key):
+        invalid_chars = '\/:*?"<>|'
+        s = ''.join(' ' if c in invalid_chars else c for c in key)
```
Comment (hanifvirani) [inline sdetools/sdelib/sqlite.py]: Use better name than ```s```. Maybe even ```c```.

File: sdetools/extlib/sqlitedict.py [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+            if not os.path.exists(dirname):
+                raise RuntimeError('Error! The directory does not exist, %s' % dirname)
+
+        self.filename = filename
+        if '"' in tablename:
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Does sqlite itself have functions for checking for invalid table names? (Is this the only invalid character?)

File: sdetools/extlib/sqlitedict.py [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+            raise ValueError('Invalid tablename %r' % tablename)
+        self.tablename = tablename
+        self.autocommit = autocommit
+        self.journal_mode = journal_mode
+
+        logger.info("opening Sqlite table %r in %s" % (tablename, filename))
+        MAKE_TABLE = 'CREATE TABLE IF NOT EXISTS "%s" (key TEXT PRIMARY KEY, value BLOB)' % self.tablename
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Should be using sqlite3's built in parameter substitution? So

```
MAKE_TABLE = 'CREATE TABLE IF NOT EXISTS "?" (key TEXT PRIMARY KEY, value BLOB)'
...
self.conn.execute(MAKE_TABLE, self.tablename)
...
```

Can you not pass parameters this way to indicate what table you want to use?

**Comment by hanifvirani** [comment]
How come we are dumping the whole sqlite file instead of using it as a dependency?

File: sdetools/extlib/sqlitedict.py [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+
+    def items(self):
+        return self.iteritems() if major_version > 2 else list(self.iteritems())
+
+    def __contains__(self, key):
+        HAS_ITEM = 'SELECT 1 FROM "%s" WHERE key = ?' % self.tablename
+        return self.conn.select_one(HAS_ITEM, (key,)) is not None
+
+    def __getitem__(self, key):
+        GET_ITEM = 'SELECT value FROM "%s" WHERE key = ?' % self.tablename
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Do we want to limit how many items we get back. We're only using the first one anyway. Do we want to raise an error if we get more than one?

File: sdetools/modules/sync_rational/rational_plugin.py (hunk 4 above)
Comment (sibartlett) [inline sdetools/modules/sync_rational/rational_plugin.py]: No, it's a defaultdict.

File: sdetools/extlib/sqlitedict.py [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+            self.update(kwds)
+
+    def __iter__(self):
+        return self.iterkeys()
+
+    def clear(self):
+        if self.flag == 'r':
+            raise RuntimeError('Refusing to clear read-only SqliteDict')
+
+        CLEAR_ALL = 'DELETE FROM "%s";' % self.tablename  # avoid VACUUM, as it gives "OperationalError: database schema has changed"
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Is `TRUNCATE TABLE blah` better / faster?

File: sdetools/extlib/sqlitedict.py [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+        else:
+            conn = sqlite3.connect(self.filename, check_same_thread=False)
+        conn.execute('PRAGMA journal_mode = %s' % self.journal_mode)
+        conn.text_factory = str
+        cursor = conn.cursor()
+        cursor.execute('PRAGMA synchronous=OFF')
+
+        res = None
+        while True:
+            req, arg, res, outer_stack = self.reqs.get()
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: Might want to comment on what messages you are expect to get in the queue?

File: sdetools/extlib/sqlitedict.py [hunk 19]
```diff
@@ -0,0 +1,520 @@
...
+        self.autocommit = autocommit
+        self.journal_mode = journal_mode
+        # use request queue of unlimited size
+        self.reqs = Queue()
+        self.setDaemon(True)  # python2.5-compatible
+        self.exception = None
+        self.log = logging.getLogger('sqlitedict.SqliteMultithread')
+        self.start()
+
+    def run(self):
```
Comment (funkaoshi) [inline sdetools/extlib/sqlitedict.py]: TODO: read this method tomorrow.

File: sdetools/modules/sync_rational/rational_plugin.py (hunk 6 above)
Comment (sibartlett) [inline sdetools/modules/sync_rational/rational_plugin.py]: We'll fix this in a separate refactor.

**Comment by sibartlett** [comment]
As mentioned in chat, sdetools can not depend on pip - so we include source for all 3rd-party dependencies (including yantra).

File: sdetools/modules/sync_rational/rational_plugin.py [hunk 20]
```diff
@@ -188,6 +193,18 @@ def initialize(self):
                                'Currently %s can only be setup as authoritative server' %
                                (self.config['conflict_policy'], self.alm_name))
 
+        self.initialize_field_cache()
+
+    def initialize_field_cache(self):
+        directory_name = '{0}__{1}'.format(
+            self.config['alm_server'].strip('/'),
+            self.config['alm_context_root'].strip('/')
+        ).replace('/', '_')
```
Comment (funkaoshi) [inline sdetools/modules/sync_rational/rational_plugin.py]: Should we slugify things? Is there anything else we need to worry about with file names?

File: sdetools/modules/sync_rational/rational_plugin.py (hunk 5 above)
Comment (funkaoshi) [inline sdetools/modules/sync_rational/rational_plugin.py]: Method might be clearer with some comments.

File: sdetools/sdelib/sqlite.py (hunk 13 above)
Comment (funkaoshi) [inline sdetools/sdelib/sqlite.py]: `s` is a string, `c` is a char? If so I think the names are fine.

File: sdetools/modules/sync_rational/rational_plugin.py (hunk 20 above)
Comment (sibartlett) [inline sdetools/modules/sync_rational/rational_plugin.py]: I was originally intending to use a slugify library, but the ones I looked at had a bunch of dependencies.

This particular line, I intend to revisit though.

File: sdetools/modules/sync_rational/rational_plugin.py (hunk 5 above)
Comment (sibartlett) [inline sdetools/modules/sync_rational/rational_plugin.py]: Done.

File: sdetools/sdelib/sqlite.py (hunk 8 above)
Comment (sibartlett) [inline sdetools/sdelib/sqlite.py]: Done.

File: sdetools/sdelib/sqlite.py (hunk 13 above)
Comment (sibartlett) [inline sdetools/sdelib/sqlite.py]: Yeah, `s` is a string and `c` is a char. I renamed `s` to `name` though.

Comment (hanifvirani) [inline sdetools/sdelib/sqlite.py]: What does the string represent though? Looks like it's a file name, so should be called ```filename``` or ```name```

File: sdetools/modules/sync_rational/rational_plugin.py (hunk 3 above)
Comment (sibartlett) [inline sdetools/modules/sync_rational/rational_plugin.py]: Done.

**Pull request merged**